gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java

```diff
@@ -16,26 +16,80 @@
 
 package com.google.gcloud;
 
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.Set;
+
 /**
  * Base class for all service exceptions.
  */
 public class BaseServiceException extends RuntimeException {
 
   private static final long serialVersionUID = 5028833760039966178L;
+  public static final int UNKNOWN_CODE = 0;
 
   private final int code;
   private final boolean retryable;
+  private final boolean idempotent;
+
+  public BaseServiceException(IOException exception, boolean idempotent) {
+    super(exception.getMessage(), exception);
+    this.code = statusCode(exception);
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(exception);
+  }
+
+  public BaseServiceException(GoogleJsonError error, boolean idempotent) {
+    super(error.getMessage());
+    this.code = error.getCode();
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(error);
+  }
 
-  public BaseServiceException(int code, String message, boolean retryable) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent) {
     super(message);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
   }
 
-  public BaseServiceException(int code, String message, boolean retryable, Exception cause) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent,
+      Exception cause) {
     super(message, cause);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
+  }
+
+  protected Set<Integer> retryableCodes() {
+    return Collections.emptySet();
+  }
+
+  protected int statusCode(IOException exception) {
+    if (exception instanceof GoogleJsonResponseException) {
```

Comment:
1) 2) and 3) What holds me from including (and relying on) the notion of *reason* is that reasons are not well documented across services. For instance there seem to be no naming convention: Datastore uses underscore separated, upper-case strings. Storage instead uses camel-case strings (not even sure about this)

1) Isn't the reason stored in `GoogleJsonError.getErrors().get(0).getReason()`?

4) This is does not seem to be what we do in `Storage`, isn't it? Do you see a benefit in doing this?

> I would still hold off in changing the Datastore api layer to throw IllegalArgumentException instead of DatastoreException and do it in a separate PR (or change the api javadoc in this one).

This PR was meant for discussion so we can trash it a start from scratch.